### PR TITLE
Add a Waged rebalancer util api that do not need raw zk address.

### PR DESCRIPTION
**Pull request by xyuanlu**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

#1755 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

ReadOnlyWagedRebalancer should not require a was Zk address in constructor.
In a multi ZK scenario, the address might be a mock address or a fake one. Passing a raw ZK address does not make sense.

### Tests

- [X] The following tests are written for this issue:
This API will be called by a REST API added in PR #1747. Unit test added in that PR will cover this change. 

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/ReadOnlyWagedRebalancer.java [hunk 1]
```diff
@@ -69,6 +71,10 @@ public void updateChangeDetectorSnapshots(ResourceControllerDataProvider dataPro
       super(new ZkBucketDataAccessor(metadataStoreAddress), clusterName);
     }
 
+    ReadOnlyAssignmentMetadataStore(ZkBucketDataAccessor zkBucketDataAccessor, String clusterName) {
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/ReadOnlyWagedRebalancer.java:74]: Optional, but if there is no other use cases, please just remove the older constructor. The newer one is more specific and avoid misusage, IMO.

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java [hunk 2]
```diff
@@ -295,7 +352,6 @@ public static String serializeByComma(List<String> objects) {
       LOG.error("getIdealAssignmentForWagedFullAuto(): Failed to compute ResourceAssignments!", e);
     } finally {
       // Close all ZK connections
-      baseDataAccessor.close();
       readOnlyWagedRebalancer.close();
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:355]: readOnlyWagedRebalancer is also passed in object, why closing it but removing baseDataAccessor close?

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java [hunk 3]
```diff
@@ -239,6 +277,25 @@ public static String serializeByComma(List<String> objects) {
         new ReadOnlyWagedRebalancer(metadataStoreAddress, globalSyncClusterConfig.getClusterName(),
             globalSyncClusterConfig.getGlobalRebalancePreference());
 
+    Map<String, ResourceAssignment> result =
+        getAssignmentForWagedFullAutoImpl(helixDataAccessor, readOnlyWagedRebalancer,
+            globalSyncClusterConfig, instanceConfigs, liveInstances, idealStates, resourceConfigs,
+            usePrefLists);
+    baseDataAccessor.close();
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:284]: close readOnlyWagedRebalancer?

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java [hunk 4]
```diff
@@ -210,19 +231,36 @@ public static String serializeByComma(List<String> objects) {
...
+    ReadOnlyWagedRebalancer readOnlyWagedRebalancer =
+        new ReadOnlyWagedRebalancer(zkBucketDataAccessor, globalSyncClusterConfig.getClusterName(),
+            globalSyncClusterConfig.getGlobalRebalancePreference());
+
+    return getAssignmentForWagedFullAutoImpl(helixDataAccessor, readOnlyWagedRebalancer,
+        globalSyncClusterConfig, instanceConfigs, liveInstances, idealStates, resourceConfigs,
+        usePrefLists);
+  }
+
+  private static Map<String, ResourceAssignment> getAssignmentForWagedFullAutoHelper(
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:259]: This method is making the code here more complicated. I suggest not adding this one but making data accessor construction logic private methods and then call them repeatedly in the existing APIs.

Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:259]: TFTR. If we make data accessor construction logic private, when adding a new REAT API, then the caller still need to pass a raw ZK address?

File: helix-core/src/main/java/org/apache/helix/util/HelixUtil.java (hunk 2 above)
Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:329]: If we pass a baseDataAccessor created by caller, then that accessor should not be closed here. readOnlyWagedRebalancer is created in this function so we close it in final block.

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:329]: For helix-reset, the caller create the accessor and passed in, right? So it is not created in this function.

Comment (xyuanlu) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:329]: Yes this is correct. So we should not close accessor here.

Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/util/HelixUtil.java:329]: I see, I was thinking about the zkBucketDataAccessor. It is not closed when the rebalancer is closed.
You are right.

File: helix-rest/src/main/java/org/apache/helix/rest/server/ServerContext.java [hunk 5]
```diff
@@ -278,6 +282,17 @@ public HelixDataAccessor getDataAccessor(String clusterName) {
     return _byteArrayZkBaseDataAccessor;
   }
 
+  public ZkBucketDataAccessor getZkBucketDataAccessor() {
```
Comment (jiajunwang) [inline helix-rest/src/main/java/org/apache/helix/rest/server/ServerContext.java:285]: Since this is not used right now, shall we put it in the next PR?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/server/ServerContext.java:285]: Yea make sense.

**Review by jiajunwang — APPROVED** [review]

**Comment by xyuanlu** [comment]
This PR is ready to be merged. Approved by @jiajunwang.

commit message.
Add a Waged rebalancer util api that do not need raw zk address

**Pull request merged**